File: docs/absorbed-from-phenotype-gfx/unity/water/src/GerstnerWaveBank.cs
using System;
using System.Collections.Generic;
using UnityEngine;

namespace Phenotype.Water
{
    /// <summary>
    /// A single Gerstner (trochoidal) wave definition.
    /// </summary>
    /// <remarks>
    /// Gerstner waves are a more realistic approximation of ocean waves than simple sine waves.
    /// They produce sharp crests and broad troughs, and displace vertices horizontally as well as vertically.
    /// </remarks>
    /// <example>
    /// <code>
    /// var wave = new GerstnerWave(
    ///     amplitude: 1.0f,
    ///     wavelength: 20.0f,
    ///     steepness: 0.5f,
    ///     direction: new Vector2(1f, 0f),
    ///     speed: 5.0f);
    /// </code>
    /// </example>
    public struct GerstnerWave
    {
        /// <summary>Crest amplitude in world units.</summary>
        /// <value>Height from water level to crest. Must be positive for visible displacement.</value>
        public float Amplitude;

        /// <summary>Distance between successive crests in world units.</summary>
        /// <value>Controls how tightly packed the wave peaks are. Smaller values = higher frequency.</value>
        public float Wavelength;

        /// <summary>
        /// Steepness (also called Q). Range [0, 1].
        /// 0 = sinusoidal; 1 = sharp trochoid with potential looping.
        /// </summary>
        /// <value>Clamped to [0, 1] in the constructor.</value>
        public float Steepness;

        /// <summary>Normalised XZ propagation direction.</summary>
        /// <value>
        /// Automatically normalised in the constructor. If a zero-length vector is supplied,
        /// falls back to <see cref="Vector2.right"/>.
        /// </value>
        public Vector2 Direction;

        /// <summary>Phase speed in world units per second.</summary>
        /// <value>How fast the wave crest moves across the surface.</value>
        public float Speed;

        /// <summary>
        /// Creates a new Gerstner wave with the specified parameters.
        /// </summary>
        /// <param name="amplitude">Crest height in world units.</param>
        /// <param name="wavelength">Crest-to-crest distance in world units.</param>
        /// <param name="steepness">Q in [0,1]. Clamped automatically.</param>
        /// <param name="direction">Propagation direction (will be normalised).</param>
        /// <param name="speed">Phase speed in world units/s.</param>
        /// <example>
        /// <code>
        /// var wave = new GerstnerWave(0.5f, 10f, 0.3f, Vector2.right, 2f);
        /// </code>
        /// </example>
        public GerstnerWave(float amplitude, float wavelength, float steepness,
                            Vector2 direction, float speed)
        {
            Amplitude = amplitude;
            Wavelength = wavelength;
            Steepness = Mathf.Clamp01(steepness);
            Direction = direction.sqrMagnitude > 1e-10f ? direction.normalized : Vector2.right;
            Speed = speed;
        }
    }

    /// <summary>
    /// Manages a bank of Gerstner wave parameters for ocean/lake surface animation.
    /// Each wave is defined by amplitude, wavelength, steepness, and direction.
    /// The bank is evaluated per-vertex to produce displaced positions and normals.
    /// </summary>
    /// <remarks>
    /// The bank is immutable after creation; use <see cref="Add"/> or the constructor
    /// that accepts an <see cref="IEnumerable{GerstnerWave}"/> to populate it.
    /// </remarks>
    /// <example>
    /// <code>
    /// var bank = GerstnerWaveBank.CreateOceanPreset();
    /// Vector3 displacement = bank.SampleDisplacement(Vector2.zero, 1.5f);
    /// Vector3 normal = bank.SampleNormal(Vector2.zero, 1.5f);
    /// </code>
    /// </example>
    public class GerstnerWaveBank
    {
        private readonly List<GerstnerWave> _waves;

        /// <summary>Read-only view of the current wave set.</summary>
        /// <value>The ordered list of waves in this bank.</value>
        public IReadOnlyList<GerstnerWave> Waves => _waves;

        /// <summary>Creates an empty bank.</summary>
        /// <example>
        /// <code>
        /// var bank = new GerstnerWaveBank();
        /// bank.Add(new GerstnerWave(1f, 10f, 0.5f, Vector2.right, 2f));
        /// </code>
        /// </example>
        public GerstnerWaveBank()
        {
            _waves = new List<GerstnerWave>();
        }

        /// <summary>
        /// Creates a bank pre-populated with <paramref name="waves"/>.
        /// </summary>
        /// <param name="waves">Initial wave collection.</param>
        /// <example>
        /// <code>
        /// var bank = new GerstnerWaveBank(new[] {
        ///     new GerstnerWave(0.5f, 20f, 0.4f, Vector2.right, 3f),
        ///     new GerstnerWave(0.3f, 8f, 0.3f, Vector2.up, 2f)
        /// });
        /// </code>
        /// </example>
        public GerstnerWaveBank(IEnumerable<GerstnerWave> waves)
        {
            _waves = new List<GerstnerWave>(waves);
        }

        /// <summary>
        /// Adds a wave to the bank and returns <c>this</c> for chaining.
        /// </summary>
        /// <param name="wave">The wave to add.</param>
        /// <returns>The current bank instance for fluent chaining.</returns>
        /// <example>
        /// <code>
        /// var bank = new GerstnerWaveBank()
        ///     .Add(new GerstnerWave(0.5f, 20f, 0.4f, Vector2.right, 3f))
        ///     .Add(new GerstnerWave(0.3f, 8f, 0.3f, Vector2.up, 2f));
        /// </code>
        /// </example>
        public GerstnerWaveBank Add(GerstnerWave wave)
        {
            _waves.Add(wave);
            return this;
        }

        // ──────────────────────────────────────────────────────────────────────
        // Core evaluation
        // ──────────────────────────────────────────────────────────────────────

        /// <summary>
        /// Sums the horizontal (XZ) and vertical (Y) Gerstner displacement at
        /// world position <paramref name="worldXZ"/> for the given <paramref name="time"/>.
        /// </summary>
        /// <param name="worldXZ">Undisplaced surface position in XZ.</param>
        /// <param name="time">Elapsed time in seconds.</param>
        /// <returns>
        /// World-space displacement vector. Add to the undisplaced position to get
        /// the final displaced position.
        /// </returns>
        /// <remarks>
        /// Waves with zero or negative amplitude or wavelength are skipped.
        /// </remarks>
        /// <example>
        /// <code>
        /// var bank = GerstnerWaveBank.CreateOceanPreset();
        /// Vector2 position = new Vector2(10f, 5f);
        /// Vector3 displacement = bank.SampleDisplacement(position, 2.5f);
        /// Vector3 worldPos = new Vector3(position.x + displacement.x, displacement.y, position.y + displacement.z);
        /// </code>
        /// </example>
        public Vector3 SampleDisplacement(Vector2 worldXZ, float time)
        {
            float dx = 0f, dy = 0f, dz = 0f;

            foreach (var w in _waves)
            {
                if (w.Amplitude <= 0f || w.Wavelength <= 0f) continue;

                float k = TwoPi / w.Wavelength;               // wave number
                float phi = w.Speed * k * time;                  // phase offset = ω·t, ω = speed·k
                float dot = k * (w.Direction.x * worldXZ.x
                               + w.Direction.y * worldXZ.y)
                           - phi;

                float sinD = Mathf.Sin(dot);
                float cosD = Mathf.Cos(dot);

                // Q·A controls horizontal crest sharpening
                float qa = w.Steepness * w.Amplitude;

                dx += -w.Direction.x * qa * sinD;
                dz += -w.Direction.y * qa * sinD;
                dy += w.Amplitude * cosD;
            }

            return new Vector3(dx, dy, dz);
        }

        /// <summary>
        /// Returns the analytic unit normal of the displaced surface at
        /// <paramref name="worldXZ"/> at the given <paramref name="time"/>.
        /// Computed from first-order partial derivatives of the displacement — no
        /// finite-difference approximation.
        /// </summary>
        /// <param name="worldXZ">Undisplaced surface position in XZ.</param>
        /// <param name="time">Elapsed time in seconds.</param>
        /// <returns>The unit normal vector at the specified position and time.</returns>
        /// <remarks>
        /// Waves with zero or negative amplitude or wavelength are skipped.
        /// If the computed normal length is near zero, returns <see cref="Vector3.up"/>.
        /// </remarks>
        /// <example>
        /// <code>
        /// var bank = GerstnerWaveBank.CreateOceanPreset();
        /// Vector2 position = new Vector2(10f, 5f);
        /// Vector3 normal = bank.SampleNormal(position, 2.5f);
        /// // Use normal for lighting or buoyancy calculations
        /// </code>
        /// </example>
        public Vector3 SampleNormal(Vector2 worldXZ, float time)
        {
            // Tangents in X and Z: dP/dx and dP/dz of the Gerstner sum.
            // Normal = cross(dP/dx, dP/dz), normalised.
            //
            // For wave i with direction d=(dx,dz), wavenumber k, amplitude A,
            // steepness Q, phase θ = k(d·p) - ω·t:
            //   ∂X/∂x = 1 - Q·A·k·dx²·cos θ   ← from horizontal displacement contribution
            //   ∂Y/∂x = A·k·dx·(-sin θ)        ← from vertical displacement contribution
            //   ∂Z/∂x = -Q·A·k·dx·dz·cos θ
            //   (and symmetrically for ∂/∂z)
            //
            // We accumulate the summed partial terms then build the cross product.

            float dXdx = 1f, dYdx = 0f, dZdx = 0f;
            float dXdz = 0f, dYdz = 0f, dZdz = 1f;

            foreach (var w in _waves)
            {
                if (w.Amplitude <= 0f || w.Wavelength <= 0f) continue;

                float k = TwoPi / w.Wavelength;
                float phi = w.Speed * k * time;
                float dot = k * (w.Direction.x * worldXZ.x
                               + w.Direction.y * worldXZ.y)
                           - phi;

                float sinD = Mathf.Sin(dot);
                float cosD = Mathf.Cos(dot);

                float wx = w.Direction.x;
                float wz = w.Direction.y;
                float qa = w.Steepness * w.Amplitude;
                float ak = w.Amplitude * k;
                float qak = qa * k;

                // ∂displacement/∂x
                dXdx -= qak * wx * wx * cosD;
                dYdx -= ak * wx * sinD;
                dZdx -= qak * wx * wz * cosD;

                // ∂displacement/∂z
                dXdz -= qak * wx * wz * cosD;
                dYdz -= ak * wz * sinD;
                dZdz -= qak * wz * wz * cosD;
            }

            // Tangent vectors in world space (dP = undisplaced + displacement)
            var tangentX = new Vector3(dXdx, dYdx, dZdx);
            var tangentZ = new Vector3(dXdz, dYdz, dZdz);

            // Normal = tangentZ × tangentX  (right-hand rule, Y-up)
            var normal = Vector3.Cross(tangentZ, tangentX);
            float len = normal.magnitude;
            return len > 1e-10f ? normal / len : Vector3.up;
        }

        // ──────────────────────────────────────────────────────────────────────
        // Factory presets
        // ──────────────────────────────────────────────────────────────────────

        /// <summary>
        /// Creates a default open-ocean preset with four varied waves covering
        /// a range of scales and directions.
        /// </summary>
        /// <returns>A pre-configured <see cref="GerstnerWaveBank"/> suitable for ocean scenes.</returns>
        /// <remarks>
        /// Wave parameters are tuned for a world-unit = 1 m scale:
        /// long swell from the south-west, mid-frequency chop from south-east
        /// and north-west, plus a short high-frequency detail wave.
        /// </remarks>
        /// <example>
        /// <code>
        /// var ocean = GerstnerWaveBank.CreateOceanPreset();
        /// Vector3 displacement = ocean.SampleDisplacement(Vector2.zero, 1f);
        /// </code>
        /// </example>
        public static GerstnerWaveBank CreateOceanPreset()
        {
            return new GerstnerWaveBank(new[]
            {
                // Primary swell — long, low-steepness, south-west to north-east
                new GerstnerWave(
                    amplitude:  0.8f,
                    wavelength: 60f,
                    steepness:  0.45f,
                    direction:  new Vector2(0.7f,  0.7f),
                    speed:      6.5f),

                // Secondary chop — shorter, south-east to north-west
                new GerstnerWave(
                    amplitude:  0.4f,
                    wavelength: 24f,
                    steepness:  0.5f,
                    direction:  new Vector2(-0.6f, 0.8f),
                    speed:      4.0f),

                // Cross-swell from north-west
                new GerstnerWave(
                    amplitude:  0.25f,
                    wavelength: 15f,
                    steepness:  0.55f,
                    direction:  new Vector2(-0.8f, -0.6f),
                    speed:      3.2f),

                // High-frequency surface ripple
                new GerstnerWave(
                    amplitude:  0.08f,
                    wavelength: 4f,
                    steepness:  0.35f,
                    direction:  new Vector2(0.4f, -0.9f),
                    speed:      2.0f),
            });
        }

        /// <summary>
        /// Creates a calm lake preset with two gentle, low-steepness waves.
        /// </summary>
        /// <returns>A pre-configured <see cref="GerstnerWaveBank"/> suitable for lake scenes.</returns>
        /// <example>
        /// <code>
        /// var lake = GerstnerWaveBank.CreateLakePreset();
        /// Vector3 displacement = lake.SampleDisplacement(Vector2.zero, 1f);
        /// </code>
        /// </example>
        public static GerstnerWaveBank CreateLakePreset()
        {
            return new GerstnerWaveBank(new[]
            {
                new GerstnerWave(
                    amplitude:  0.05f,
                    wavelength: 10f,
                    steepness:  0.2f,
                    direction:  new Vector2(1f, 0.3f),
                    speed:      1.5f),

                new GerstnerWave(
                    amplitude:  0.03f,
                    wavelength: 6f,
                    steepness:  0.15f,
                    direction:  new Vector2(-0.5f, 1f),
                    speed:      1.0f),
            });
        }

        // ──────────────────────────────────────────────────────────────────────
        // Helpers
        // ──────────────────────────────────────────────────────────────────────

        private const float TwoPi = 2f * Mathf.PI;
    }
}

File: docs/absorbed-from-phenotype-gfx/src/water/gerstner_wave_bank.rs
//! Gerstner (trochoidal) wave bank.
//!
//! Ported from C# `GerstnerWaveBank.cs` + `GerstnerWave` struct. Gerstner
//! waves are a more realistic approximation of ocean waves than simple sine
//! waves. They produce sharp crests and broad troughs, and displace vertices
//! horizontally as well as vertically.

use std::f32::consts::PI;

use crate::water::error::WaterError;

const TWO_PI: f32 = 2.0 * PI;
const NORMAL_EPSILON: f32 = 1e-10;

/// A single Gerstner (trochoidal) wave definition.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct GerstnerWave {
    /// Crest amplitude in world units.
    pub amplitude: f32,
    /// Distance between successive crests in world units.
    pub wavelength: f32,
    /// Steepness (Q). Range `[0, 1]`. 0 = sinusoidal, 1 = sharp trochoid.
    pub steepness: f32,
    /// Normalised XZ propagation direction. (x, z) tuple.
    pub direction: (f32, f32),
    /// Phase speed in world units per second.
    pub speed: f32,
}

impl GerstnerWave {
    /// Create a new Gerstner wave.
    ///
    /// `steepness` is clamped to `[0, 1]`. `direction` is normalised; if a
    /// zero-length direction is supplied, falls back to `(1, 0)` (east).
    pub fn new(amplitude: f32, wavelength: f32, steepness: f32, direction: (f32, f32), speed: f32) -> Self {
        let steepness = steepness.clamp(0.0, 1.0);
        let mag2 = direction.0 * direction.0 + direction.1 * direction.1;
        let direction = if mag2 > 1e-10 {
            let inv_mag = 1.0 / mag2.sqrt();
            (direction.0 * inv_mag, direction.1 * inv_mag)
        } else {
            (1.0, 0.0)
        };
        Self { amplitude, wavelength, steepness, direction, speed }
    }
}

/// Manages a bank of Gerstner wave parameters for ocean/lake surface animation.
#[derive(Debug, Clone, Default)]
pub struct GerstnerWaveBank {
    waves: Vec<GerstnerWave>,
}

impl GerstnerWaveBank {
    /// Create an empty bank.
    pub fn new() -> Self { Self::default() }

    /// Create a bank pre-populated with `waves`.
    pub fn from_waves(waves: Vec<GerstnerWave>) -> Self {
        Self { waves }
    }

    /// Read-only view of the current wave set.
    pub fn waves(&self) -> &[GerstnerWave] { &self.waves }

    /// Number of waves in the bank.
    pub fn len(&self) -> usize { self.waves.len() }

    /// Whether the bank has no waves.
    pub fn is_empty(&self) -> bool { self.waves.is_empty() }

    /// Adds a wave to the bank and returns `self` for chaining.
    pub fn add(&mut self, wave: GerstnerWave) -> &mut Self {
        self.waves.push(wave);
        self
    }

    /// Sums the horizontal (XZ) and vertical (Y) Gerstner displacement at
    /// `world_xz` for the given `time`.
    pub fn sample_displacement(&self, world_xz: glam::Vec2, time: f32) -> glam::Vec3 {
        let mut dx = 0.0f32;
        let mut dy = 0.0f32;
        let mut dz = 0.0f32;

        for w in &self.waves {
            if w.amplitude <= 0.0 || w.wavelength <= 0.0 { continue; }

            let k = TWO_PI / w.wavelength;
            let phi = w.speed * k * time;
            let dot = k * (w.direction.0 * world_xz.x + w.direction.1 * world_xz.y) - phi;
            let sin_d = dot.sin();
            let cos_d = dot.cos();
            let qa = w.steepness * w.amplitude;

            dx += -w.direction.0 * qa * sin_d;
            dz += -w.direction.1 * qa * sin_d;
            dy += w.amplitude * cos_d;
        }

        glam::Vec3::new(dx, dy, dz)
    }

    /// Returns the analytic unit normal of the displaced surface at
    /// `world_xz` at the given `time`. If the normal length is near zero,
    /// returns `glam::Vec3::Y`.
    pub fn sample_normal(&self, world_xz: glam::Vec2, time: f32) -> glam::Vec3 {
        let mut d_xdx = 1.0f32;
        let mut d_ydx = 0.0f32;
        let mut d_zdx = 0.0f32;
        let mut d_xdz = 0.0f32;
        let mut d_ydz = 0.0f32;
        let mut d_zdz = 1.0f32;

        for w in &self.waves {
            if w.amplitude <= 0.0 || w.wavelength <= 0.0 { continue; }

            let k = TWO_PI / w.wavelength;
            let phi = w.speed * k * time;
            let dot = k * (w.direction.0 * world_xz.x + w.direction.1 * world_xz.y) - phi;
            let sin_d = dot.sin();
            let cos_d = dot.cos();
            let wx = w.direction.0;
            let wz = w.direction.1;
            let qa = w.steepness * w.amplitude;
            let ak = w.amplitude * k;
            let qak = qa * k;

            d_xdx -= qak * wx * wx * cos_d;
            d_ydx -= ak * wx * sin_d;
            d_zdx -= qak * wx * wz * cos_d;
            d_xdz -= qak * wx * wz * cos_d;
            d_ydz -= ak * wz * sin_d;
            d_zdz -= qak * wz * wz * cos_d;
        }

        let tangent_x = glam::Vec3::new(d_xdx, d_ydx, d_zdx);
        let tangent_z = glam::Vec3::new(d_xdz, d_ydz, d_zdz);
        let normal = tangent_z.cross(tangent_x);
        if normal.length() > NORMAL_EPSILON { normal.normalize() } else { glam::Vec3::Y }
    }

    /// Create a default open-ocean preset with four varied waves covering a
    /// range of scales and directions.
    pub fn create_ocean_preset() -> Self {
        Self::from_waves(vec![
            GerstnerWave::new(0.8, 60.0, 0.45, (0.7,  0.7), 6.5),
            GerstnerWave::new(0.4, 24.0, 0.5,  (-0.6, 0.8), 4.0),
            GerstnerWave::new(0.25, 15.0, 0.55, (-0.8, -0.6), 3.2),
            GerstnerWave::new(0.08, 4.0, 0.35, (0.4, -0.9), 2.0),
        ])
    }

    /// Create a calm lake preset with two gentle, low-steepness waves.
    pub fn create_lake_preset() -> Self {
        Self::from_waves(vec![
            GerstnerWave::new(0.05, 10.0, 0.2,  (1.0, 0.3),  1.5),
            GerstnerWave::new(0.03,  6.0, 0.15, (-0.5, 1.0), 1.0),
        ])
    }
}

/// Validate a `time` value. Negative times are allowed in the simulation
/// (waves can be evaluated at t=0 in the past) but `NaN` / `±∞` is not.
pub fn validate_time(time: f32) -> Result<(), WaterError> {
    if time.is_nan() || time.is_infinite() {
        return Err(WaterError::OutOfBounds {
            msg: format!("time must be finite, got {}", time),
        });
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn approx(a: f32, b: f32, tol: f32) -> bool { (a - b).abs() < tol }

    #[test]
    fn empty_bank_has_no_waves() {
        let b = GerstnerWaveBank::new();
        assert!(b.is_empty());
        assert_eq!(b.len(), 0);
    }

    #[test]
    fn add_waves_grows_bank() {
        let mut b = GerstnerWaveBank::new();
        b.add(GerstnerWave::new(1.0, 10.0, 0.5, (1.0, 0.0), 2.0));
        b.add(GerstnerWave::new(0.5, 5.0, 0.3, (0.0, 1.0), 1.0));
        assert_eq!(b.len(), 2);
    }

    #[test]
    fn empty_bank_displacement_is_zero() {
        let b = GerstnerWaveBank::new();
        let d = b.sample_displacement(glam::Vec2::ZERO, 1.0);
        assert!(approx(d.x, 0.0, 1e-6));
        assert!(approx(d.y, 0.0, 1e-6));
        assert!(approx(d.z, 0.0, 1e-6));
    }

    #[test]
    fn empty_bank_normal_is_unit_y() {
        let b = GerstnerWaveBank::new();
        let n = b.sample_normal(glam::Vec2::ZERO, 1.0);
        assert!(approx(n.x, 0.0, 1e-6));
        assert!(approx(n.y, 1.0, 1e-6));
        assert!(approx(n.z, 0.0, 1e-6));
    }

    #[test]
    fn ocean_preset_has_four_waves() {
        assert_eq!(GerstnerWaveBank::create_ocean_preset().len(), 4);
    }

    #[test]
    fn lake_preset_has_two_waves() {
        assert_eq!(GerstnerWaveBank::create_lake_preset().len(), 2);
    }

    #[test]
    fn ocean_preset_normal_is_unit_length() {
        let b = GerstnerWaveBank::create_ocean_preset();
        let n = b.sample_normal(glam::Vec2::new(7.3, -4.1), 2.5);
        assert!(approx(n.length(), 1.0, 1e-4));
    }

    #[test]
    fn wave_with_zero_amplitude_is_skipped() {
        let mut b = GerstnerWaveBank::new();
        b.add(GerstnerWave::new(0.0, 10.0, 0.5, (1.0, 0.0), 1.0));
        b.add(GerstnerWave::new(1.0, 10.0, 0.5, (0.0, 1.0), 1.0));
        let d = b.sample_displacement(glam::Vec2::ZERO, 0.0);
        // Only the second wave contributes; we just check it's not zero.
        assert!(d.y.abs() > 0.0);
    }

    #[test]
    fn zero_direction_falls_back_to_east() {
        let w = GerstnerWave::new(1.0, 10.0, 0.5, (0.0, 0.0), 1.0);
        assert_eq!(w.direction, (1.0, 0.0));
    }

    #[test]
    fn steepness_is_clamped() {
        let w = GerstnerWave::new(1.0, 10.0, 5.0, (1.0, 0.0), 1.0);
        assert_eq!(w.steepness, 1.0);
        let w = GerstnerWave::new(1.0, 10.0, -1.0, (1.0, 0.0), 1.0);
        assert_eq!(w.steepness, 0.0);
    }

    #[test]
    fn validate_time_rejects_nan() {
        assert!(validate_time(f32::NAN).is_err());
    }
}
